File: algo-java/Tree94.java
import java.util.*;

public class Tree94 {

    //输入：root = [1,null,2,3]
    //输出：[1,3,2]

    class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode() {
        }

        TreeNode(int val) {
            this.val = val;
        }

        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }
    List<Integer> res =new ArrayList<>();
    class Solution {
        public List<Integer> inorderTraversal(TreeNode root) {



            //边界
            if (root == null) return null;
            //左 中 右子树

            if(root.left!=null){

                inorderTraversal(root.left);

            }
            res.add(root.val);
            if(root.right!=null){
                inorderTraversal(root.right);
            }

            return res;
        }

        public List<Integer> inorderTraversal2(TreeNode root) {

            Stack<TreeNode> stack  = new Stack();
            TreeNode tmp = root;
            while(!stack.isEmpty() || root!=null){
                if(root!=null){
                    stack.add(tmp);
                    tmp= tmp.left;
                }else {
                    tmp = stack.pop();
                    res.add(tmp.val);
                    root = tmp.right;
                }


            }

            return res;
        }
    }
}

File: algo/Tree94.py
# This is a sample Python script.

# Press ⌃R to execute it or replace it with your code.
# Press Double ⇧ to search everywhere for classes, files, tool windows, actions, and settings.



class TreeNode(object):
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class Solution(object):
    def inorderTraversal(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """

        res = []
        s= []
        p=root
        while p or s:
            if p :
                s.append(p)
                p=p.left
            else:
                tmp = s.pop()
                res.append(tmp.val)
                p = tmp.right

        return res
